src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -10,19 +12,26 @@ internal partial class LazyItemEvaluator<P, I, M, D>
     {
         class RemoveOperation : LazyItemOperation
         {
+            readonly ImmutableList<string> _matchOnMetadata;
+
             public RemoveOperation(OperationBuilder builder, LazyItemEvaluator<P, I, M, D> lazyEvaluator)
                 : base(builder, lazyEvaluator)
             {
+                _matchOnMetadata = builder.ItemElement.MatchOnMetadata.Equals(string.Empty) ? null : builder.ItemElement.MatchOnMetadata.Split(';').ToImmutableList();
             }
 
             // todo port the self referencing matching optimization (e.g. <I Remove="@(I)">) from Update to Remove as well. Ideally make one mechanism for both. https://github.com/Microsoft/msbuild/issues/2314
             // todo Perf: do not match against the globs: https://github.com/Microsoft/msbuild/issues/2329
             protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+                if (_matchOnMetadata != null && (_itemSpec.Fragments.Count != 1 || _itemSpec.ItemSpecString.Contains('%')))
```

Comment:
I think you can copy paste and adjust the full check from here: https://github.com/microsoft/msbuild/pull/4997/files#diff-0d4693ff9ac2fd56e6fda64d0fd0c6cbR275-R281